pkg/controller/workspacerouting/workspacerouting_controller.go

```diff
@@ -175,33 +176,95 @@ func (r *ReconcileWorkspaceRouting) Reconcile(request reconcile.Request) (reconc
+func (r *ReconcileWorkspaceRouting) reconcileStatus(
+		instance *workspacev1alpha1.WorkspaceRouting,
+		routingObjects solvers.RoutingObjects,
+		exposedEndpoints map[string][]workspacev1alpha1.ExposedEndpoint) error {
 	if instance.Status.Phase == workspacev1alpha1.RoutingReady &&
-		cmp.Equal(instance.Status.PodAdditions, routingObjects.PodAdditions) &&
-		cmp.Equal(instance.Status.ExposedEndpoints, routingObjects.ExposedEndpoints) {
+			cmp.Equal(instance.Status.PodAdditions, routingObjects.PodAdditions) &&
+			cmp.Equal(instance.Status.ExposedEndpoints, exposedEndpoints) {
 		return nil
 	}
 	instance.Status.Phase = workspacev1alpha1.RoutingReady
 	instance.Status.PodAdditions = routingObjects.PodAdditions
-	instance.Status.ExposedEndpoints = routingObjects.ExposedEndpoints
+	instance.Status.ExposedEndpoints = exposedEndpoints
 	return r.client.Status().Update(context.TODO(), instance)
 }
 
+func getExposedEndpoints(
+		endpoints map[string][]workspacev1alpha1.Endpoint,
+		ingresses []v1beta1.Ingress,
+		routes []routeV1.Route) (map[string][]workspacev1alpha1.ExposedEndpoint, error) {
+	exposedEndpoints := map[string][]workspacev1alpha1.ExposedEndpoint{}
+
+	for machineName, machineEndpoints := range endpoints {
+		for _, endpoint := range machineEndpoints {
+			if endpoint.Attributes[workspacev1alpha1.PUBLIC_ENDPOINT_ATTRIBUTE] != "true" {
+				continue
+			}
+			url, err := getURLforEndpoint(endpoint, ingresses, routes)
+			if err != nil {
+				return nil, err
+			}
+			exposedEndpoints[machineName] = append(exposedEndpoints[machineName], workspacev1alpha1.ExposedEndpoint{
+				Name:       endpoint.Name,
+				Url:        url,
+				Attributes: endpoint.Attributes,
+			})
+		}
+	}
+	return exposedEndpoints, nil
+}
+
+func getURLforEndpoint(
+		endpoint workspacev1alpha1.Endpoint,
+		ingresses []v1beta1.Ingress,
+		routes []routeV1.Route) (string, error) {
+	endpointName := common.EndpointName(endpoint.Name)
+	for _, route := range routes {
+		if route.Labels[config.WorkspaceEndpointNameLabel] == endpointName {
+			protocol := endpoint.Attributes[workspacev1alpha1.PROTOCOL_ENDPOINT_ATTRIBUTE]
+			if endpoint.Attributes[workspacev1alpha1.SECURE_ENDPOINT_ATTRIBUTE] == "true" &&
+					route.Spec.TLS != nil {
+				protocol = getSecureProtocol(protocol)
+			}
+			url := fmt.Sprintf("%s://%s", protocol, route.Spec.Host)
+			return url, nil
+		}
+	}
+	for _, ingress := range ingresses {
+		if ingress.Labels[config.WorkspaceEndpointNameLabel] == endpointName {
+			if len(ingress.Spec.Rules) == 1 {
+				protocol := endpoint.Attributes[workspacev1alpha1.PROTOCOL_ENDPOINT_ATTRIBUTE]
+				url := fmt.Sprintf("%s://%s", protocol, ingress.Spec.Rules[0].Host)
+				return url, nil
+			} else {
+				return "", fmt.Errorf("ingress %s contains multiple rules", ingress.Name)
+			}
+		}
+	}
+	return "", fmt.Errorf("could not get URL for endpoint %s", endpoint.Name)
```

Comment:
Is it really an error? While workspace is starting I got quite a lot of spam
![Screenshot_20200414_152116](https://user-images.githubusercontent.com/5887312/79224454-c4fd0600-7e63-11ea-857a-81acf57142ed.png)

Are we able to return only that endpoints which are already exposed and schedule requeuing? 
other (without ingress/route) could be skipped or returned with status - Exposing (not sure if we have endpoint status now).

Reply:
Makes sense -- where are you running the operator? It may be a route/ingress exposure issue (i.e. it takes longer than expected). I don't see this on `crc`.

I'll add this logic to the PR.

Reply:
Looking into this a bit more, I'm not sure what's causing the issue; we only return that error when we cannot find any ingress with the appropriate label. If the url field is empty, we just return empty.

Reply:
I tried it on local crc

Reply:
> Looking into this a bit more, I'm not sure what's causing the issue; we only return that error when we cannot find any ingress with the appropriate label

Exactly. Now I see the same error but only once
![route-not-found](https://user-images.githubusercontent.com/5887312/79327783-b541f800-7f1d-11ea-8dca-b77d808f4103.gif)

So, instead of failing reconciling loop it would be better to make endpoint as problematic without host and retry later. I think it worths the dedicated PR because it needs introducing some phases,conditions for endpoints.

Reply:
This error also appears after removing the workspacerouting route![Screenshot_20200415_133739](https://user-images.githubusercontent.com/5887312/79328179-4ca74b00-7f1e-11ea-8716-b69a45fa41b5.png)

Reply:
@sleshchenko I think the error you describe initially (`could not get URL...`) *is* a failure; we should never reach that point in the reconcile and not have routes/ingresses available. The previous step, prior to attempting this matching, is to make sure routes/ingresses are in sync between the cluster and spec. Not being able to match the two means that we have an endpoint (which defines the spec), that does not have a route/ingress associated with it. Until we have a concrete case where this can occur, it should mark the workspace as failed.

The latter failure, `already exists`, is a kind of familiar issue -- I haven't gotten around to implementing it, but we shouldn't log that error and just continue. It just means the state of the cluster has changed since we started our reconcile loop. The correct solution there would be to just requeue if `errors.IsConflict(err)` or `errors.IsAlreadyExists(err)`

Reply:
> Exactly. Now I see the same error but only once

This is a different error message, to be clear.

Reply:
> This is a different error message, to be clear.

My fault. But seems there were two tries to create ingress and it can indicate in the issues in the reconciling loop, because I did not create such a route but hand.

Will test more precisely. BTW it's not a blocker

Reply:
Yeah, it's `already exists` issue and seems routes are not created immediately 
![Screenshot_20200416_151950](https://user-images.githubusercontent.com/5887312/79455973-61efa880-7ff6-11ea-8334-78239c8bf2f7.png)

RetryAfter 200 helps me to solve an issue. 100 is not enough... 
```go
	routesInSync, clusterRoutes, err := r.syncRoutes(instance, routes)
	if err != nil || !routesInSync {
		reqLogger.Info("Routes not in sync")
		return reconcile.Result{ RequeueAfter: 100 * time.Millisecond}, err
	}
```
But not sure if it's a really good solution since this limit depends on the infrastructure, I'm OK with leaving this error propagated constantly for time being.

Reply:
The main thing stopping us implementing this is that it's going to be a fair bit of boilerplate error checking; it might also be improved by eventually filtering reconciles (by determining if changes are necessary)